common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java

```diff
@@ -374,6 +378,30 @@ private void applyAfterEvaluate(final Project project) {
                     definitionSet.forEach((identifier, definition) -> {
                         definition.configureRun(runImpl);
                     });
+
+                    //Handle dev login.
+                    final DevLogin devLogin = project.getExtensions().getByType(Subsystems.class).getDevLogin();
+                    final Tools tools = project.getExtensions().getByType(Subsystems.class).getTools();
+                    if (devLogin.getEnabled().get()) {
+                        //Dev login is only supported on the client side
+                        if (runImpl.getIsClient().get() && runImpl.getShouldUseDevLogin().get()) {
+                            final String mainClass = runImpl.getMainClass().get();
+
+                            //We add the dev login tool to the runtime only configuration, of the first source set, this should suffice
+                            final SourceSet defaultSourceSet = runImpl.getModSources().get().get(0);
+                            final Configuration defaultRuntimeOnlyConfiguration = project.getConfigurations().maybeCreate(defaultSourceSet.getRuntimeOnlyConfigurationName());
+                            defaultRuntimeOnlyConfiguration.getDependencies().add(project.getDependencies().create(tools.getDevLogin().get()));
+
+                            //Update the program arguments to properly launch the dev login tool
+                            run.getProgramArguments().add("--launch_target");
+                            run.getProgramArguments().add(mainClass);
+
+                            //Set the main class to the dev login tool
+                            run.getMainClass().set("net.covers1624.devlogin.DevLogin");
+                        } else if (!runImpl.getIsClient().get() && runImpl.getShouldUseDevLogin().get()) {
+                            throw new GradleException("Dev login is only supported on runs which are marked as clients! The run: " + runImpl.getName() + " is not a client run.");
```

Comment:
Perhaps [`InvalidUserDataException`](https://docs.gradle.org/8.7/javadoc/org/gradle/api/InvalidUserDataException.html) would be a more fitting exception here, than the generic `GradleException`.